### ISSUE TITLE
Question about permissions on Lua side

### ISSUE THREAD

**Issue by ricanteja** [ThePhD/sol2#254]
First of all thank you for this awesome library!

My question is this: How do I make a table/namespace and its content readonly/const?
Let's say I have a class
```cpp
struct Object
{
    void myFunc()
    {
        cout << "hello\n";
```
and then I bind it in...
```cpp
int main()
{
    sol::state lua;
    lua.open_libraries(sol::lib::base);

    Object my_obj;
    
    sol::table obj_table = lua.create_named_table("object");
    obj_table.set_function("my_func", sol::readonly(&Object::myFunc), &my_obj);
```
So that my interface on the lua side will look like this
```lua
object.my_func()
```
 I read in the docs that if I used `sol::readonly()` that the value should not be overwritten and that is really the behavior I want for the whole table and its member functions. There should only exist 1 instance of Object during the whole execution of the program. I don't want the user to be able to create new `Object` but I have some problems with this idea.

How do I stop the end user from totally destroying the lua interface by doing this?:
```lua
object = nil
```
Also I tested and `sol::readonly()` doesn't do what I expected. I am able to do this:
```lua
object.my_func = nil
```

Basically I'm trying to create a programming interface by exporting some functionality to lua but I want to be able to block the user from overwriting these tables and their functionality.

**Comment by ThePhD** [comment]
So, in order to make a regular table readonly you need to hijack the `__newindex` metamethod of a table's metatable and set it to nil. `sol::readonly` and those things only work for usertypes, because we explicitly control the metatable in those cases.

Granted, this is easy for you to do with Sol2. Here's a complete example:

````
#define SOL_CHECK_ARGUMENTS
#include <sol.hpp>

#include <iostream>

struct Object {
	void myFunc() {
		std::cout << "hello\n";
	}
};

int deny(lua_State* L) {
	return luaL_error(L, "HAH! Deniiiiied!");
}

int main() {
	sol::state lua;
	lua.open_libraries(sol::lib::base);

	Object my_obj;

	sol::table obj_table = lua.create_named_table("object");
	
	sol::table obj_metatable = lua.create_table_with();
	obj_metatable.set_function("my_func", &Object::myFunc, &my_obj);
	// Set whatever else you need to	
	
	// Properly self-index metatable to block things
	obj_metatable[sol::meta_function::new_index] = deny;
	obj_metatable[sol::meta_function::index] = obj_metatable;
	
	// Set it on the actual table
	obj_table[sol::metatable_key] = obj_metatable;
	
	try {
		lua.script(R"(
print(object.my_func)
object["my_func"] = 24
print(object.my_func)
		)");
	}
	catch (const std::exception& e) {
		std::cout << e.what() << std::endl;
	}
	return 0;
}
````

Of course, I say "easy", but a person who doesn't know how Lua works underneath probably couldn't write this code. Therefore, I might have to open an issue request to further generalize the purpose of `sol::readonly`, or to perhaps have a method on tables called `my_table.make_readonly();`. Still, the above code should get you on your way.

Good luck.
